compiler/src/iree/compiler/PluginAPI/PluginManager.h

```diff
@@ -0,0 +1,233 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <optional>
+#include <string_view>
+#include <vector>
+
+#include "iree/compiler/Utils/OptionUtils.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringMap.h"
+
+namespace mlir {
+class DialectRegistry;
+class MLIRContext;
+}  // namespace mlir
+
+namespace mlir::iree_compiler {
+
+class AbstractPluginSession;
+class PluginManager;
+struct PluginManagerSession;
+class PluginRegistrar;
+
+// Registration functions are exported with this signature.
+using PluginRegistrationFunction = bool (*)(PluginRegistrar *);
+
+// Empty options class that satisfies the contract for an OptionsBinder.
+// Used by default if a plugin does not support options.
+struct EmptyPluginOptions {
+  static void bindOptions(OptionsBinder &binder) {}
+};
+
+// Abstract class representing a plugin registration. It is responsible for
+// various global initialization and creation of plugin sessions that mirror
+// the lifetime of and |iree_compiler_session_t| for when the plugin is
+// activated.
+//
+// This is typically not instantiated directly but via the PluginSession
+// CRTP helper which manages most details.
+class AbstractPluginRegistration {
+ public:
+  AbstractPluginRegistration(std::string pluginId)
+      : pluginId(std::move(pluginId)) {}
+  virtual ~AbstractPluginRegistration();
+
+  // Gets the plugin id. Valid for the life of the registration.
+  std::string_view getPluginId() { return pluginId; }
+
+  // Performs once-only global initialization. This is called prior to any
+  // sessions being created and affects everything in the process. It is
+  // best used for passes and other such things.
```

Comment:
Could you perhaps give an example? At this point it's unclear why best for passes or what one would do with pass during init